movici_simulation_core/models/rules/model.py

```diff
@@ -0,0 +1,425 @@
+        """Load rules from config or rules dataset.
+
+        :param init_data_handler: Handler for loading datasets
+        :returns: Rules specification dict with 'defaults' and 'rules' keys
+        :rtype: dict
+        """
+        if "rules" in self.config:
+            return {"rules": self.config["rules"], "defaults": self.config.get("defaults", {})}
+
+        if "rules_dataset" in self.config:
+            dataset_name = self.config["rules_dataset"]
+            _, path = init_data_handler.get(dataset_name)
+            if path is None:
+                raise ValueError(f"Rules dataset '{dataset_name}' not found")
+            data = path.read_dict()
+            return data.get("data", {})
+
+        return {"rules": [], "defaults": {}}
+
+    def _register_rule_attributes(
+        self,
+        state: TrackedState,
+        schema: AttributeSchema,
+        init_data_handler: InitDataHandler,
+    ) -> None:
+        """Register source and target attributes for all rules.
+
+        :param state: TrackedState instance
+        :param schema: AttributeSchema for attribute specs
+        :param init_data_handler: Handler for loading entity data
+        """
+        # Track registered datasets/entity groups
+        registered: dict[tuple[str, str], t.Any] = {}
+
+        for rule in self.rules:
+            # Register source attributes if needed
+            attr_names = rule.condition.get_attribute_names()
+            if attr_names and rule.from_dataset:
+                self._register_source_attributes(
+                    rule, attr_names, state, schema, init_data_handler, registered
+                )
+
+            # Register target attribute
+            if rule.to_dataset and rule.output:
+                self._register_target_attribute(rule, state, schema, init_data_handler, registered)
+
+    def _register_source_attributes(
+        self,
+        rule: Rule,
+        attr_names: set[str],
+        state: TrackedState,
+        schema: AttributeSchema,
+        init_data_handler: InitDataHandler,
+        registered: dict,
+    ) -> None:
+        """Register source attributes for a rule's condition.
+
+        :param rule: The rule being processed
+        :param attr_names: Set of attribute names referenced in the condition
+        :param state: TrackedState instance
+        :param schema: AttributeSchema
+        :param init_data_handler: Handler for loading entity data
+        :param registered: Dict tracking registered entity groups
+        """
+        # Find which entity group contains the source entity
+        entity_group_name = self._find_entity_group(
+            rule.from_dataset,
+            rule.from_id,
+            rule.from_reference,
+            init_data_handler,
+        )
+
+        if entity_group_name is None:
+            if self.logger:
+                self.logger.warning("Could not find entity group for source entity in rule")
```

Comment:
Might be helpful to also output in which dataset you're looking and what is the ref/id of the entity in case of an error